### DISCUSSION TITLE
Multiple uploads cause endpoint to hang (asyncio.exceptions.CancelledError)

### DISCUSSION THREAD

**Discussion by guites** [fastapi/fastapi#8346] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, UploadFile, File
import uvicorn
from fastapi.responses import HTMLResponse

app = FastAPI()

@app.get("/")
async def main():
    content = """
        <body>
            <form method="post">
                <input id="fileinput" name="file" type="file" multiple>
            </form>
            <script>
                const input = document.getElementById('fileinput');
                
                const upload = (file) => {
                const formData = new FormData();
                formData.append('file', file);
                fetch('/test', {
                    method: 'POST',
                    body: formData
                }).then(
                    response => response.json() // if the response is a JSON object
                ).then(
                    success => console.log(success) // Handle the success response object
                ).catch(
                    error => console.log(error) // Handle the error response object
                );
                };

                const onSelectFile = () => {
                    for (var i = 0; i < 100; i++) {
                        upload(input.files[0]);
                    }
                }

                input.addEventListener('change', onSelectFile, false);
            </script>
        </body>
    """
    return HTMLResponse(content=content)


@app.post("/test")
async def test_upload(file: UploadFile = File(...)):
    return {"file_name": file.filename}
    

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8000)
```

### Description

This is a rather strange bug, but the code above seems to be the minimal way in which I could reproduce it.

What happens is, whenever there is an endpoint that accepts a file, if you make enough concurrent requests with a file larger than about ~60kb, a small part of these requests will hang.

In the "/", I'm serving a simple upload form, which sends the file 100 times via javascript fetch to the "/test" endpoint, which is a reproduction of the file upload method present in the tutorial ([here](https://fastapi.tiangolo.com/tutorial/request-files/#file-parameters-with-uploadfile)).

Here is a screen grab of what happens:
![hang_requests](https://user-images.githubusercontent.com/71985299/178852639-e1204c0e-c188-41da-a026-5a2c02fa5aa7.png)
The first four requests listed will stay that way indefinitely.

If I then go to my terminal and press `ctrl + c`, I will get the following error:

> INFO:     127.0.0.1:42746 - "POST /test HTTP/1.1" 200 OK
^CINFO:     Shutting down
INFO:     Waiting for connections to close. (CTRL+C to force quit)
^CINFO:     Finished server process [16982]
ERROR:    Traceback (most recent call last):
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/routing.py", line 638, in lifespan
    await receive()
  File "/home/vscode/.local/lib/python3.9/site-packages/uvicorn/lifespan/on.py", line 137, in receive
    return await self.receive_queue.get()
  File "/usr/local/lib/python3.9/asyncio/queues.py", line 166, in get
    await getter
asyncio.exceptions.CancelledError
>
>ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/vscode/.local/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 403, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/vscode/.local/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/home/vscode/.local/lib/python3.9/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/home/vscode/.local/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/routing.py", line 266, in handle
    await self.app(scope, receive, send)
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/routing.py", line 65, in app
    response = await func(request)
  File "/home/vscode/.local/lib/python3.9/site-packages/fastapi/routing.py", line 192, in app
    body = await request.form()
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/requests.py", line 254, in form
    self._form = await multipart_parser.parse()
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/formparsers.py", line 190, in parse
    async for chunk in self.stream:
  File "/home/vscode/.local/lib/python3.9/site-packages/starlette/requests.py", line 219, in stream
    message = await self._receive()
  File "/home/vscode/.local/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 540, in receive
    await self.message_event.wait()
  File "/usr/local/lib/python3.9/asyncio/locks.py", line 226, in wait
    await fut
asyncio.exceptions.CancelledError
INFO:     127.0.0.1:42752 - "POST /test HTTP/1.1" 500 Internal Server Error

I tried handling the file uploads within a `try - except - finally` block, but the error persists:

```python
@app.post("/test")
async def test_upload(file: UploadFile = File(...)):
    try:
        await file.read()
    except Exception as e:
        print(str(e))
    finally:
        await file.close()
```

If I make these request in a "syncronous" fashion, for example, looping over a directory filled with images and then using `curl` on each one, the problem will not occur.

I'm sorry if I'm missing something obvious. Thanks.

### Operating System

Linux

### Operating System Details

I am running the project on a docker container. It can be cloned from [this repository](https://github.com/guites/fast-api-upload-issue) (contains only the minimal parts to reproduct issue).

### FastAPI Version

0.78.0

### Python Version

3.9.10

**Comment by iudeen** [comment]
Is it waiting indefinitely?

**Comment by guites** [comment]
> Is it waiting indefinitely?

Yes, it seems to hang indefinitely.

edit: the requests eventually error out after 10min with http 500. What is strange is that the terminal of my fastAPI app does not log any 500 status.

![error_10min](https://user-images.githubusercontent.com/71985299/178996874-fa6bcb0a-8ca4-4556-b053-1b13405487a8.png)

**Comment by iudeen** [comment]
I am not able to reproduce, it works fine for me. The requests are handled in chucks. But eventually, all the 100 gets a success response.

**Comment by guites** [comment]
> I am not able to reproduce, it works fine for me. The requests are handled in chucks. But eventually, all the 100 gets a success response.

Thanks. I still have the problem with the pending requests, here is a screenshot on chrome:

![pending_requests](https://user-images.githubusercontent.com/71985299/178992952-6d3e15a7-3782-4816-ae72-14cf05dd694f.png)

I've found these two stackoverflow posts which report pending requests, though they seem to be about doing heavy computation instead of handling files: https://stackoverflow.com/questions/71966225/fastapi-request-from-client-side-stays-pending https://stackoverflow.com/questions/67599119/fastapi-asynchronous-background-tasks-blocks-other-requests

**Comment by iudeen** [comment]
The SO links are different. In their case they have synchronous function which is bound to block the execution. But you don't have a sync logic. 

What version of FastAPI and Unicorn are you running? Can you update to latest and give a try?

**Comment by Kokopak** [comment]
I had the same problem with a logging middleware that I put in my code. Maybe you got something like this

**Comment by bsilverthorn** [comment]
I can repro this problem consistently using @guites' script, but for some reason only when I'm hitting it through VS Code port forwarding. Haven't yet been able to repro when running the script locally, even with browser network throttling, or when forwarding via pure `SSH -L ...`.

Presumably a timing sensitivity but :shrug:.

**Comment by guites** [comment]
@bsilverthorn Yeah I was using VS Code's devcontainer when the bug originally occurred.

I will try and reproduce this on a different machine, and without using VSCode on the original ubuntu machine, and report back.

edit: Yeah can confirm the bug only happens from VS Code devcontainer. Using [this](https://github.com/guites/fast-api-upload-issue/blob/main/Dockerfile) Dockerfile and running

```
docker build . -t bug-repro-fastapi
docker run -p 8000:8000 bug-repro-fastapi
```

the bug doesn't reproduces.

**Comment by ZdsAlpha** [comment]
I am facing the same issue with VSCode and fastapi. It only doesn't work for VS Code port forwarding.
